### DISCUSSION TITLE
Questions about more complex scenarios

### DISCUSSION THREAD

**Discussion by frederikhors** [SeaQL/sea-orm#1139] [Q&A]
I have a project with these models and relations:

1. Player
    1. Team
        1. Games
        1. Coach
        1. Doctor
            1. MedicalSpecialty
    1. Sponsor

This is an example, of course, but as you can see I can have on table `Player` a `team_id` column as FK (Foreign Key) for `Team` which has `coach_id` column as FK for `Coach` table.

In my Golang backend I'm using [an ORM](https://github.com/ent/ent) that allows you to (generate and) use this kind of code:

```go
type Player struct {
  id string
  name string
  Team *Team
  //...
}

type Team struct {
  id string
  Coach *Coach
  //...
}

type Coach struct {
  id string
  //...
}

func (repo *Repo) PlayerById(id string) DomainPlayer {
  player := repo.queryById(id).withTeam().withCoach()

  // player here has TEAM and TEAM.COACH using a unique SQL query

  // and I can easily convert it:

  return ConvertPlayerToDomainPlayer(player)
}
```

Starting with Rust I was fascinated by [SeaORM](https://github.com/SeaQL/sea-orm) but now that I'm trying to create something more complex I have realized that I cannot do things like in Go.

I'm [generating entities using `sea-orm-cli`](https://www.sea-ql.org/SeaORM/docs/generate-entity/sea-orm-cli/) from my Postgresql DB created using [SeaORM Migrator](https://www.sea-ql.org/SeaORM/docs/migration/writing-migration/) (this is totally optional, you can also write the following code by yourself).

The generated entities are like this:

```rust
//! SeaORM Entity. Generated by sea-orm-codegen 0.9.3

use sea_orm::entity::prelude::*;

#[derive(Clone, Debug, PartialEq, DeriveEntityModel)]
#[sea_orm(table_name = "player")]
pub struct Model {
    #[sea_orm(primary_key, unique)]
    pub id: String,
    pub name: String,
    pub team_id: Option<String>,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(
        belongs_to = "super::team::Entity",
        from = "Column::TeamId",
        to = "super::team::Column::Id",
        on_update = "NoAction",
        on_delete = "NoAction"
    )]
    Team,
}

impl Related<super::team::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Team.def()
    }
}

impl ActiveModelBehavior for ActiveModel {}
```

And I'm using code like:

```rust
async fn player_by_id(&self, id: String) -> Result<Option<DomainPlayer>> {
    let player = match Player::find_by_id(id)
        .find_also_related(Team)
        // I would like to use here something like `with_team()` and `with_coach()` or `with_team(|team| team.with_coach())`
        .one(self.db)
        .await?
    {
        Some(o) => o,
        None => return Ok(None),
    };

    let coach = if let Some(team) = &player.1 {
        Coach::find_by_id(team.coach_id).one(self.db).await?
    } else {
        None
    };

    // Since I need to return a DomainPlayer (not a Player) here I cannot use a `Player.into()` because generated `Player` does not have `Team` field

    Ok(Some(custom_func_to_transform_player_to_domain_player(player, player.1 /*this is team*/, coach)))

    // Do you can imagine how complex is to use these functions?
}
```

**I'd like to know what you think of this mental order.**

Most likely I need to change it because the code is really ugly (as well as expensive for the many SQL queries and allocations) and inconvenient to use to scale with even more complexity.

**What do you recommend?**

**Comment by ikrivosheev** [comment]
@frederikhors hello! I think you need something like this: https://github.com/SeaQL/sea-orm/issues/739

**Reply by frederikhors** [reply]
Thank you very much for the reply.

I've read a bit but can't figure out if it's really what I'm looking for.

What do you use now?

**Reply by ikrivosheev** [reply]
The idea is: you write some struct, derive and you can convert SQL rows into struct. Without a lot of manual written code. 

At the moment I have a lot of manual  written code...

**Reply by frederikhors** [reply]
It would be a dream! I found https://cetra3.github.io/blog/implementing-a-jobq-sqlx. I don't know sqlx, but maybe this is already doable using `#[derive(sqlx::FromRow)]`?

**Reply by ikrivosheev** [reply]
Sqlx can not aggregate by some key (I mean relationship). Also it can't inline struct, for example, you have fields: **image.size**, **image.weight**. You want to convert fields: image.* Into inner struct.

**Comment by billy1624** [comment]
Hey @frederikhors, I think this article explains the limitation of Rust regarding selecting complex relations.
- https://www.sea-ql.org/sea-orm-cookbook/001-eager-load-nested-relation.html

We're considering a data loader approach in SeaORM which is capable to select nested relation in batch.

CC @tyt2y3

**Reply by frederikhors** [reply]
> We're considering a data loader approach in SeaORM which is capable to select nested relation in batch.

Is there any issue to follow?

**Reply by billy1624** [reply]
No, I don't think so

**Reply by frederikhors** [reply]
This is crucial for us. I think they can force me to drop seaorm in favour of pure sqlx. I think we need an issue for this.

**Reply by frederikhors** [reply]
@tyt2y3 is there an issue for this today to follow?

**Reply by tyt2y3** [reply]
I think work is being done on #1191
